tests/testthat/test-tcplhit2_core.R

```diff
@@ -13,3 +13,150 @@ test_that("tcplhit2 works", {
   expect_equal(output$tp, 0.749, tolerance = 1e-2)
   expect_equal(output$ga, 9.59, tolerance = 1e-2)
 })
+
+test_that("tcplhit2 BMD boundary check", {
+  # Preparation for tests on BMD boundary arguments
+  # Simulate some data
+  X <- rep(seq(0.001,3,length.out = 10),each = 5)
+
+  set.seed(918)
+  Case_1 <- hillfn(ps = c(4,0.00105,1.08,0.1),x = X) + rt(n = length(X),df = 4)
+  Case_3 <- pow(ps = c(10,5,0.1),x = X) + rt(n = length(X),df = 4)
+  set.seed(455)
+  Case_2 <- hillfn(ps = c(4,0.00105,1.08,0.1),x = X) + rt(n = length(X),df = 4)
+  set.seed(321)
+  Case_4 <- exp2(ps = c(5.6,15,0.1),x = X) + rt(n = length(X),df = 4)
+  set.seed(322)
+  Case_5 <- exp5(ps = c(4,10,5,0.1),x = X) + rt(n = length(X),df = 4)
+
+  Y <- rbind(Case_1, Case_2, Case_3, Case_4, Case_5)
+
+  # Y contains 5 rows of simulated responses, each row corresponds to
+  # one data cases in order:
+
+  # 1. BMD < lower threshold
+  # 2. lower threshold < BMD < lowest expt dose
+  # 3. lowest expt dose < BMD < upper expt dose
+  # 4. upper expt dose < BMD < upper threshold
+  # 5. BMD > upper threshold
+
+  df <- matrix(nrow = 5, ncol = 14)
+  colnames(df) <- c("onesd", "cutoff", "bmd", "bmdu", "bmdl",
+                    "combo1", "combo1u", "combo1l",
+                    "combo2", "combo2u", "combo2l",
+                    "combo3", "combo3u", "combo3l")
+
+  # fit each data case and hit-call with different argument combinations
+  for (i in 1:5) {
+    temp <- Y[i, 1:10]
+    bmad <- mad(temp)
```

Comment:
BMAD should be mad(Y[,1:10]), that is the median absolute deviation of replicates in the two lowest concentrations across all responses - NOT the median absolute deviation of replicates in the two lowest concentrations for a single response.

The bmad and the cutoff can be calculated prior to the for loop.  The cutoff assignment to the 'df' matrix can be done prior to the for loop as well.

Reply:
@sedavid01 Y is a 5x50 matrix, each row is one data case (10 doses, 5 replicates each). I thought each data case should be independent and calculate bmed and cutoff separately. `Y[i, 1:10]` are replicates in the two lowest concentrations for the i-th data case. Let me know if I misunderstood.

Reply:
Each row in your Y is an individual response (i.e. data case or response case).

- onesd - calculated on the individual response (single row :: single response)
- bmad & cutoff - across a group (or all) of responses in the dataset (all rows :: all responses)